smtp_client.js

```diff
@@ -256,6 +273,52 @@ class SMTPClient extends events.EventEmitter {
         this.send_command('RSET');
     }
 
+    shutdown () {
+        if (!this.connected || this.command === 'data' || this.command === 'mailbody') {
+            // Destroy here, we can't reuse a connection that was mid-data.
+            this.destroy();
+            return;
+        }
+
+        logger.logdebug(`[smtp_client_pool] ${this.uuid} shutting down, state=${this.state}`);
+        if (this.state === STATE.DESTROYED) {
+            return;
+        }
+
+        this.state = STATE.SHUTDOWN;
+        this.removeAllListeners('greeting');
+        this.removeAllListeners('capabilities');
+        this.removeAllListeners('xclient');
+        this.removeAllListeners('helo');
+        this.removeAllListeners('mail');
+        this.removeAllListeners('rcpt');
+        this.removeAllListeners('data');
+        this.removeAllListeners('dot');
+        this.removeAllListeners('rset');
+        this.removeAllListeners('auth');
+        this.removeAllListeners('error');
+        this.removeAllListeners('bad_code');
+        this.removeAllListeners('rset');
```

Comment:
most of this `shutdown()` is a copy/paste of `release()`. Except the errors, like listing 'rset' twice. Not only should it be refactored to avoid duplicate code, but it should also be refactored for hygiene. Something like this is a fair bit cleaner IMO:

```js
['greeting', 'capabilities', 'xclient', 'helo', 'mail', 'rcpt', 'data', 'dot', 'rset', 'auth', 'error', 'bad_code'].forEach(l => {
    this.removeAllListeners(l);
})
```